### PR TITLE
Update to handle multiple inputs

### LINKED ISSUE #1: Cannot pass multiple inputs to a model
At present the library only allows us to pass a single tensor to a model and receive a single output tensor.

This needs updating so that we can pass multiple input tensors.

In the Fortran:
- Instead of passing an input tensor pass an array of tensors.
- Get the length of the input array of tensors and pass that to the c++

In the c++:
- loop over the number of tensors and append (`.push_back`) on to the inputs to the C function.

### PR DESCRIPTION
**Pull request by jatkinson1000**
These updates should address and close #1 

In the Fortran:
- [x] Instead of passing an input tensor pass an array of tensors.
- [x] Get the length of the input array of tensors and pass that to the c++

In the c++:
- [x] loop over the number of tensors and append (`.push_back`) on to the inputs to the C function.
- [x] Work out how output can be cast back to array of pointers.

In the examples:
- [x] Update so that they still compile and run (make inputs/outputs arrays, and add n_in/out)

In the documentation:
- [x] Update documentation to reflect these changes (ONLY CHANGES. OTHER DOCS ISSUES ARE FOR ANOTHER ISSUE!)

### CODE REVIEW

File: fortran-pytorch-lib/ts_inference.c [hunk 1]
```diff
@@ -39,14 +39,16 @@ int main(int argc, const char* argv[])
   float* output_data = (float*)malloc(output_size * sizeof(float));
 
   if (model) {
-    torch_tensor_t input = torch_from_blob(input_data, input_ndim, input_shape,
+    torch_tensor_t inputs[1];
+    inputs[0] = torch_from_blob(input_data, input_ndim, input_shape,
                                            torch_kFloat32, torch_kCPU);
+    const int nin = 1;
```
Comment (mondus) [inline fortran-pytorch-lib/ts_inference.c:45]: It seems at odds to make this change and then not have a test using multiple inputs. The correct thing to do is to have some unit testing (but that is probably another issue). There should at least be an example for now that allows testing using multiple inputs.

**Review by mondus — CHANGES_REQUESTED** [review]
Some more minor comments. Testing is a suggestion to protect your future self so I would be happy to leave that to discretion. Take a look at the try catch loop though.

**Review by SimonClifford — APPROVED** [review]
lgtm

File: fortran-pytorch-lib/ctorch.cpp [hunk 2]
```diff
@@ -166,15 +166,21 @@ torch_jit_script_module_t torch_jit_load(const char* filename)
...
-  inputs.push_back(*in);
+  // Generate a vector of IValues (placeholders for various Torch types)
+  std::vector<torch::jit::IValue> inputs_vec;
+  // Populate with Tensors pointed at by pointers
+  for (int i=0; i<nin; ++i) {
+    inputs_vec.push_back(*(in[i]));
+  }
   try {
-    std::move(*out) = model->forward(inputs).toTensor();
+    std::move(*out) = model->forward(inputs_vec).toTensor();
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply fortran-pytorch-lib/ctorch.cpp:183]: We are casting the output of calling the model to a tensor here, and catching if something goes wrong, so wouldn't be able to test using `isTensor` until after we perform this operation.
As such I think this is OK as is, but willing to discuss in-person and probe further.

File: fortran-pytorch-lib/ctorch.cpp [hunk 3]
```diff
@@ -183,6 +189,7 @@ void torch_jit_module_forward(const torch_jit_script_module_t module,
     exit(EXIT_FAILURE);
   }
   // FIXME: this should be the responsibility of the user
+  // TODO: perform this check on all elements(or first and alter rest)?
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply fortran-pytorch-lib/ctorch.cpp:192]: Removing this line as we are currently only dealing with a single output.
We will need to revisit this in future when looking at multiple outputs in issue #3 , perhaps in a future release?

File: fortran-pytorch-lib/ctorch.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply fortran-pytorch-lib/ctorch.cpp:183]: See example here: https://github.com/pytorch/pytorch/blob/e380d692dc2c20d711c2155fc580eec30383d301/aten/src/ATen/core/ivalue.h#LL1083C1-L1086C8

We want to check the object we are going to cast is a tensor before doing so. Then modify the else clause to provide something like:
```
std::cerr << "[ERROR]: " << "function torch.jit.module.forward called returning object that is not a tensor" << std::endl;
```

[the remark this answers is not in the document]
Comment (jatkinson1000) [reply fortran-pytorch-lib/ctorch.cpp:183]: The documentation for `isTensor` is [here](https://pytorch.org/cppdocs/api/structc10_1_1_i_value.html#_CPPv4NK3c106IValue8isTensorEv) and based on trying to call it from a `torch::Tensor` using, e.g.
```
torch::Tensor LocalTensor;
LocalTensor = *(in[i]);
if (LocalTensor.isTensor()){
...
```
Throws an error with `error: ‘class torch::nn::AnyValue’ has no member named ‘isTensor’` so perhaps it needs calling on an `IValue` or `AnyValue` to see if iot is a Tensor.

Further, in line 174 when we do a `reinterpret_cast` on `inputs` aren't we forcing the code to view `in` as a vector of pointers to Tensors?

Another issue with implementing what we discussed is that I can't work out how `AnyValue`s should be declared. I've tried `torch::AnyValue` and `torch::jit::AnyValue`...

[the remark this answers is not in the document]
Comment (jatkinson1000) [reply fortran-pytorch-lib/ctorch.cpp:195]: OK, we now check the `IValues` to ensure that the inputs are all Tensors.
For the output we agreed that there should be a C++ error raised when we try and place them in a Tensor variable if they are not Tensors - a comment has been made to note this in the code.

We may revisit this if we look at multiple outputs in #3

**Comment by jatkinson1000** [comment]
Have addressed the above review with collaboration with @mondus and additional documentation.
Requesting re-review.

**Pull request merged**